File: ios/Curio/Curio/Intelligence/EmbeddingPreference.swift
import Foundation

/// Which backend computes embeddings. Mirrors `EmbeddingBackend` / `EmbeddingPreference` from
/// `data/embedding/EmbeddingPreference.kt`.
enum EmbeddingBackend: String, CaseIterable, Sendable {
    /// On-device EmbeddingGemma when downloaded, else the xAI cloud fallback.
    case auto = "AUTO"
    /// Force on-device EmbeddingGemma; never falls back to the cloud.
    case onDevice = "ON_DEVICE"
    /// Force the xAI cloud embedder; never runs on-device.
    case xai = "XAI"

    var settingsLabel: String {
        switch self {
        case .auto: return "Auto"
        case .onDevice: return "On-device"
        case .xai: return "xAI"
        }
    }

    var settingsDescription: String {
        switch self {
        case .auto:
            return "On-device when the model is downloaded, otherwise xAI cloud."
        case .onDevice:
            return "Private, on-device only. Requires the EmbeddingGemma model (download below)."
        case .xai:
            return "xAI cloud. Note: xAI has no public embeddings endpoint yet, so this often returns nothing."
        }
    }
}

/// Persists the chosen `EmbeddingBackend`. Uses the same UserDefaults key as Android
/// (`embedding_backend` in `curio_embedding_prefs`).
enum EmbeddingPreference {
    private static let backendKey = "embedding_backend"

    static func get() -> EmbeddingBackend {
        guard let raw = UserDefaults.standard.string(forKey: backendKey),
              let backend = EmbeddingBackend(rawValue: raw) else {
            return .auto
        }
        return backend
    }

    static func set(_ backend: EmbeddingBackend) {
        UserDefaults.standard.set(backend.rawValue, forKey: backendKey)
    }
}

File: app/src/main/java/com/example/data/embedding/EmbeddingPreference.kt
package com.example.data.embedding

import android.content.Context

/**
 * Which backend computes embeddings. The user can override the automatic choice from Settings so a
 * "0 embedded" result is diagnosable: e.g. forcing [ON_DEVICE] makes it obvious the local model
 * isn't downloaded, and forcing [XAI] surfaces that xAI ships no public embeddings endpoint yet
 * (see [com.example.data.remote.GrokModels.EMBEDDING]).
 */
enum class EmbeddingBackend {
    /** On-device EmbeddingGemma when downloaded, else the xAI cloud fallback. Default. */
    AUTO,

    /** Force on-device EmbeddingGemma; never falls back to the cloud (fully private). */
    ON_DEVICE,

    /** Force the xAI cloud embedder; never runs on-device. */
    XAI
}

/**
 * Persists the chosen [EmbeddingBackend]. Shares the same prefs file as
 * [com.example.background.EmbeddingIndexScheduler] so all embedding settings live together.
 */
object EmbeddingPreference {

    private const val PREFS = "curio_embedding_prefs"
    private const val KEY_BACKEND = "embedding_backend"

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun get(context: Context): EmbeddingBackend =
        prefs(context).getString(KEY_BACKEND, null)
            ?.let { name -> runCatching { EmbeddingBackend.valueOf(name) }.getOrNull() }
            ?: EmbeddingBackend.AUTO

    fun set(context: Context, backend: EmbeddingBackend) {
        prefs(context).edit().putString(KEY_BACKEND, backend.name).apply()
    }
}
